src/Parser/Template.cs

```diff
@@ -54,6 +57,95 @@ public Template(string template, bool isPositional, List<Literal> literals, List
       Holes = holes.ToArray();
     }
 
+    public string Render(object[] parameters)
+    {
+        var sb = new StringBuilder();
+        int pos = 0;
+        int h = 0;
+        foreach (var literal in Literals)
+        {
+            sb.Append(Value, pos, literal.Print);
+            pos += literal.Print;
+            if (literal.Skip == 0) 
+                pos++;
+            else
+            {
+                pos += literal.Skip;
+                if (IsPositional)
+                    RenderHolePositional(sb, ref Holes[h++], parameters);
+                else
+                    RenderHole(sb, ref Holes[h], parameters[h++]);
+            }
+        }
+        return sb.ToString();
+    }
+
+    private static void RenderHolePositional(StringBuilder sb, ref Hole hole, object[] parameters)
+        => RenderHole(sb, ref hole, parameters[hole.Index], true);
+
+    private static void RenderHole(StringBuilder sb, ref Hole hole, object value, bool legacy = false)
+    {
+        // TODO: handle value == null
+        
+        // TODO: destructuring {@x}
+        
+        if (hole.CaptureType == CaptureType.Stringification)
+        {
+            // TODO: we don't need to support format and alignment here?
+            sb.Append('"').Append(value.ToString()).Append('"');
+            return;
+        }
+
+        // Shortcut common case. It is important to do this before IEnumerable, as string _is_ IEnumerable
+        if (value is string)
+        {
+            AppendValue(sb, ref hole, value, legacy);
+            return;
+        }
+
+        IEnumerable collection;
+        if (!legacy && (collection = value as IEnumerable) != null)
+        {
+            int pos = sb.Length;
+            foreach (var item in collection) {
+                AppendValue(sb, ref hole, item, false);
+                sb.Append(", ");
+            }
+            if (sb.Length > pos) sb.Length -= 2; // Remove trailing ", "
```

Comment:
I was curious so I did a test. This is using a simple array of short strings (basically the numbers 1 to 50), so it puts emphasis on the condition itself, because the amount of work in the loop is tiny. Even so, at 50 elements both methods have the same timing, within a stddev:

 Method |      Median |     StdDev |                                                                                       
------- |------------ |----------- |                                                                                       
 Remove | 502.7678 ns | 12.1957 ns |                                                                                       
     If | 507.0890 ns | 11.9343 ns |                                                                                       
   Join | 726.9758 ns | 19.4694 ns | 

I expected `Join` to perform significantly better, especially since it accepts a `string[]` overload rather than `IEnumerable<string>` (where the interface dispatch would have explained the diff). Maybe poor buffer size choice explains the difference? (I made my `StringBuilder` big enough from the start in all cases).

Conclusion: we should do the simplest. Also, the remove version has a _tiny_ chance of overflowing needlessly on the last append, so it's probably better to use the flag.